app/src/main/java/org/p2p/wallet/tokenservice/TokenService.kt

```diff
@@ -0,0 +1,65 @@
+package org.p2p.wallet.tokenservice
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.launch
+import org.p2p.core.common.di.AppScope
+import org.p2p.core.token.Token
+import org.p2p.wallet.bridge.interactor.EthereumInteractor
+import org.p2p.wallet.home.events.EthereumTokensLoader
+import org.p2p.wallet.home.events.SolanaTokensLoader
+import org.p2p.wallet.user.interactor.UserTokensInteractor
+
+class TokenService(
+    private val userTokensInteractor: UserTokensInteractor,
+    private val ethereumInteractor: EthereumInteractor,
+    private val solanaTokensLoader: SolanaTokensLoader,
+    private val ethereumTokensLoader: EthereumTokensLoader,
+    private val appScope: AppScope,
+    private val loadStateHelper: TokenServiceLoadStateHelper
+) {
+
+    private val tokensState = MutableStateFlow<TokenState>(TokenState.Idle)
+
+    init {
+        appScope.launch {
+            userTokensInteractor.getUserTokensFlow().combine(
+                ethereumInteractor.observeTokensFlow()
+            ) { solTokens, ethTokens ->
+                mapToTokenState(solTokens, ethTokens)
+            }.collect(tokensState)
+        }
+    }
+
+    fun observeUserTokens(): Flow<TokenState> {
+        return tokensState
+    }
+
+    fun onStart() {
+        appScope.launch {
+            solanaTokensLoader.onStart()
+            ethereumTokensLoader.onStart().takeIf { ethereumTokensLoader.isEnabled() }
```

Comment:
`takeIf` is about mapping, not launching smth. 

```
if (ethereumTokensLoader.isEnabled()) {
      ethereumTokensLoader.onStart()
}
```
Just make it simple, there is no need to use complex extensions